### PR TITLE
Introducing OSX Support!

### PR DESCRIPTION
**Pull request by joshgarde**
Compiles qemu quite happily on my machine. Had to disable userspace emulation which currently isn't available on OSX.

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -55,7 +77,7 @@ esac
 
 function rmStep(){
   rm -rf "$@"
-  rmdir -p --ignore-fail-on-non-empty `dirname "$@"`
+  #rmdir -p --ignore-fail-on-non-empty `dirname "$@"`
```
Comment (joshgarde) [inline scripts/build]: Also, not quite sure what this is used for when `rm -rf` is used above it.

**Comment by joshgarde** [comment]
Build from my machine:
[mac-x64.tar.gz](https://github.com/NodeOS/qemu/files/333522/mac-x64.tar.gz)

**Comment by piranna** [comment]
That's awesome! :-D There are some tidy bits, besides that the only problem is SemaphoreCI only have Linux buildchains. I think Travis has OSX ones, could you be able to check and port to it?

File: scripts/build [hunk 2]
```diff
@@ -39,12 +37,36 @@ esac
...
-  'Linux')     PLATFORM=linux   ;;
-  'FreeBSD')   PLATFORM=freebsd ;;
-  'WindowsNT') PLATFORM=win     ;;
-  'Darwin')    PLATFORM=mac     ;;
-  'SunOS')     PLATFORM=solaris ;;
-  'AIX')       PLATFORM=linux   ;;
+  'Linux')
+    PLATFORM=linux
+    AUDIO=alsa
+    TARGET_LIST=arm-softmmu,i386-softmmu,x86_64-softmmu,arm-linux-user,i386-linux-user,x86_64-linux-user
```
Comment (piranna) [inline scripts/build]: Use a global TARGET_LIST and just add here the user space ones.

File: scripts/build (hunk 1 above)
Comment (piranna) [inline scripts/build]: It's used to remove empty folders.

File: scripts/build [hunk 3]
```diff
@@ -72,7 +94,7 @@ function err(){
 SRC_DIR=`pwd`/deps/qemu
 OBJ_DIR=build/$MACHINE
 OUT_DIR=`pwd`
-PRODUCTS=(bin libexec share)
+PRODUCTS=$OS_PRODUCTS
```
Comment (piranna) [inline scripts/build]: Set PRODUCTS directly on the OS selector before.

File: scripts/build [hunk 4]
```diff
@@ -116,7 +139,7 @@ if [[ ! -d $STEP_DIR ]]; then
                        --disable-snappy           \
                        --disable-vnc              \
                        --disable-xen              \
-                       --enable-sdl               || exit 6
+                       --enable-$DISPLAY               || exit 6
```
Comment (piranna) [inline scripts/build]: Align the 'exit' keyword.

File: scripts/build (hunk 1 above)
Comment (joshgarde) [inline scripts/build]: Odd - Haven't had to use rmdir right after a `rm -rf`. Gives errors on OSX related to the `--ignore-fail-on-non-empty` flag

Comment (piranna) [inline scripts/build]: > > +  #rmdir -p --ignore-fail-on-non-empty `dirname "$@"`
>
> Odd - Haven't had to use rmdir right after a rm -rf. Gives errors on OSX
related to the --ignore-fail-on-non-empty flag

It's just for clean-up and tidy-ness purposzez. What kind of errors are you
getting? This flag prevent to show error messages (and error status? Don't
remember...) when it get a not empty folder. Probably we could find a
workaround.

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+osx_image: xcode7.3
+language: node_js
+node_js:
+  - "4.4.x"
+script:
+  - npm install
```
Comment (piranna) [inline .travis.yml]: I'm not sure if Linux builds should be generated on Travis too to have them integrated... What do you think?

Comment (joshgarde) [inline .travis.yml]: It's useful for me to see what builds break on what operating system all in one place.

Comment (piranna) [inline .travis.yml]: > It's useful for me to see what builds break on what operating system all
in one place.

https://travis-ci.org/NodeOS/qemu/pull_requests

Abdolutely :-D

File: scripts/build [hunk 6]
```diff
@@ -39,12 +38,35 @@ esac
...
+    PLATFORM=freebsd
+    ;;
+  'WindowsNT')
+    PLATFORM=win
+    ;;
+  'Darwin')
+    PLATFORM=mac
+    AUDIO=coreaudio
+    DISPLAY=cocoa
+    PRODUCTS=(bin share)
```
Comment (piranna) [inline scripts/build]: Is 'libexec' not generated on OSX?

Comment (joshgarde) [inline scripts/build]: Nope - libexec contains the network bridge helper, which doesn't show up for OSX.

**Comment by joshgarde** [comment]
Travis port is finalized. 😄

**Comment by piranna** [comment]
It was tought, eh? ;-)

**Comment by piranna** [comment]
Why it's being used both brew and macports?

**Comment by joshgarde** [comment]
Travis uses Brew to handle it's dependencies because they don't have a MacPorts image.
I prefer using MacPorts, but for the sake of compile times, I bodged it a bit.

**Pull request merged**